pkg/queue/concurrency_state.go

```diff
@@ -0,0 +1,135 @@
+*/
+
+package queue
+
+import (
+	"bytes"
+	"fmt"
+	"net/http"
+
+	"go.uber.org/zap"
+)
+
+// ConcurrencyStateHandler tracks the in flight requests for the pod. When the requests
+// drop to zero, it runs the `pause` function, and when requests scale up from zero, it
+// runs the `resume` function. If either of `pause` or `resume` are not passed, it runs
+// the respective local function(s). The local functions are the expected behavior; the
+// function parameters are enabled primarily for testing purposes.
+func ConcurrencyStateHandler(h http.Handler, logger *zap.SugaredLogger, pause, resume func(string) error, endpoint string) http.HandlerFunc {
+	logger.Info("Concurrency state endpoint set, tracking request counts, using endpoint ", endpoint)
+
+	type req struct {
+		w http.ResponseWriter
+		r *http.Request
+
+		done chan struct{}
+	}
+
+	reqCh := make(chan req)
+	doneCh := make(chan struct{})
+	go func() {
+		inFlight := 0
+
+		// This loop is entirely synchronous, so there's no cleverness needed in
+		// ensuring open and close dont run at the same time etc. Only the
+		// delegated ServeHTTP is done in a goroutine.
+		for {
+			select {
+			case <-doneCh:
+				inFlight--
+				if inFlight == 0 {
+					logger.Info("Requests dropped to zero")
+					err := pause(endpoint)
+					if err != nil {
+						panic(err)
+					}
+					logger.Info("Request successfully processed")
+				}
+
+			case r := <-reqCh:
+				inFlight++
+				if inFlight == 1 {
+					logger.Info("Requests increased from zero")
+					err := resume(endpoint)
+					if err != nil {
+						panic(err)
+					}
+					logger.Info("Request successfully processed")
+				}
+
+				go func(r req) {
+					h.ServeHTTP(r.w, r.r)
+					close(r.done) // Return from ServeHTTP
+					doneCh <- struct{}{}
+				}(r)
+			}
+		}
+	}()
+
+	return func(w http.ResponseWriter, r *http.Request) {
+		done := make(chan struct{})
+		reqCh <- req{w, r, done}
+		// Block till we've processed the request
+		<-done
+	}
+}
+
+// Pause sends to an endpoint when request counts drop to zero
+func Pause(endpoint string) error {
+	action := ConcurrencyStateMessageBody{Action: "pause"}
+	bodyText := []byte(`{"action": "` + action.Action + `"}`)
+	body := bytes.NewBuffer(bodyText)
+	req, err := http.NewRequest(http.MethodPost, endpoint, body)
+	if err != nil {
+		return fmt.Errorf("unable to create request: %w", err)
+	}
+	req.Header.Add("Token", "nil") // TODO: use serviceaccountToken from projected volume
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return fmt.Errorf("unable to post request: %w", err)
+	}
+	if resp.StatusCode != http.StatusOK {
+		return fmt.Errorf("expected 200 response, got: %d: %s", resp.StatusCode, resp.Status)
+	}
+
+	return nil
+}
+
+// Resume sends to an endpoint when request counts increase from zero
+func Resume(endpoint string) error {
+	action := ConcurrencyStateMessageBody{Action: "resume"}
+	bodyText := []byte(`{"action": "` + action.Action + `"}`)
```

Comment:
I don't understand why we're building this with string concatenation when we've created a struct with json tags on the line above? Should this be `json.Marshal(action)` instead?